src/class/macros.rs

```diff
@@ -3,51 +3,46 @@
 #[macro_export]
 #[doc(hidden)]
 macro_rules! py_unary_func {
-    ($trait:ident, $class:ident :: $f:ident, $conv: expr) => {
-        py_unary_func!($trait, $class::$f, $conv, *mut $crate::ffi::PyObject);
-    };
-    // Use call_ref! by default
-    ($trait:ident, $class:ident :: $f:ident, $conv: expr, $ret_type:ty) => {
-        py_unary_func!(
-            $trait,
-            $class::$f,
-            $conv,
-            $ret_type,
-            call_ref_with_converter
-        );
-    };
-    ($trait: ident,
-     $class:ident :: $f:ident,
-     $conv: expr,
-     $ret_type: ty,
-     $call: ident
-    ) => {{
+    ($trait: ident, $class:ident :: $f:ident, $call:ident, $ret_type: ty  $(, $conv:expr)?) => {{
         unsafe extern "C" fn wrap<T>(slf: *mut $crate::ffi::PyObject) -> $ret_type
         where
             T: for<'p> $trait<'p>,
         {
-            let py = $crate::Python::assume_gil_acquired();
-            let _pool = $crate::GILPool::new(py);
-            let slf = py.from_borrowed_ptr::<$crate::PyCell<T>>(slf);
-            $call!(slf, $conv, py, $f)
+            let gil = $crate::assume_gil_acquired();
+            let py = gil.py();
+            $crate::catch_unwind!(py, {
+                let _pool = $crate::GILPool::new(py);
+                let slf = py.from_borrowed_ptr::<$crate::PyCell<T>>(slf);
+                $crate::callback::convert(py, $call!(slf, $f)$(.map($conv))?)
```

Comment:
Why this `map` works?
Isn't `conv` struct?

Reply:
I'm assuming that `conv` is a function which takes one argument and returns something new.

It just so happens that tuple structs with only one parameter can also behave as such functions:
E.g.

```
struct Foo<T>(pub T);

fn wrap(x: Option<i32>) -> Foo<i32> {
     x.map(Foo)
}
```